File: 백준/Gold/2206. 벽 부수고 이동하기/벽 부수고 이동하기.cc
#include <bits/stdc++.h>
using namespace std;

struct Point {
	int y;
	int x;
	int wall;
};

int a[1000][1000];
bool visited[3][1000][1000];
int dy[4] = {1, -1 , 0, 0};
int dx[4] = {0, 0, 1, -1};

bool isValidRange(int ny, int nx, int wall, int n, int m) {
	if (ny < 0 || nx < 0 || ny >= n || nx >= m) return false;
	if (visited[wall][ny][nx]) return false;
	return true;
}

int bfs(int n, int m) {
	queue<Point> q;
	q.push({0, 0, 0});
	visited[0][0][0] = true;
	int result = 0;
	while(!q.empty()) {
		int size = q.size();
		result++;
		while(size--) {
			Point p = q.front();
			int y = p.y;
			int x = p.x;
			int wall = p.wall;
			q.pop();
			if (y == n-1 && x == m-1) return result;
			
			for (int dir = 0; dir < 4; dir++) {
				int ny = y + dy[dir];
				int nx = x + dx[dir];
				if (!isValidRange(ny, nx, wall, n, m)) continue;
				
				visited[wall][ny][nx] = true;
				
				if (wall == 0 && a[ny][nx] == 1) {
					q.push({ny, nx, wall+1});
				} 
				else if (a[ny][nx] == 0) {
					q.push({ny, nx, wall});
				}
				
			}
		}
	}
	
	return -1;
}

int main() {
	int n, m;
	cin >> n >> m;
	for (int i = 0; i < n; i++) {
		string s;
		cin >> s;
		for (int j = 0; j < s.size(); j++) {
			a[i][j] = s[j]-'0';
		}
	}

	cout << bfs(n, m);
	return 0;
}

File: 백준/Gold/2206. 벽 부수고 이동하기/벽 부수고 이동하기.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.LinkedList;
import java.util.Queue;
import java.util.StringTokenizer;

public class Main {

    static int n, m;
    static int[][] a;
    static int[] dy = {1, -1, 0, 0};
    static int[] dx = {0, 0, -1, 1};

    static class Node {
        int y;
        int x;
        int breakCount;

        Node (int y, int x, int breakCount) {
            this.y = y;
            this.x = x;
            this.breakCount = breakCount;
        }
    }

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());
        n = Integer.parseInt(st.nextToken());
        m = Integer.parseInt(st.nextToken());
        a = new int[n][m];
        for (int i = 0; i < n; i++) {
            String temp = br.readLine();
            for (int j = 0; j < temp.length(); j++) {
                a[i][j] = temp.charAt(j) - '0';
            }
        }

        System.out.println(bfs());
    }

    static int bfs() {
        Queue<Node> q = new LinkedList<>();
        boolean[][][] visited = new boolean[n][m][2];
        q.offer(new Node(0, 0, 0));
        visited[0][0][0] = true;

        int result = 0;
        while (!q.isEmpty()) {
            int size = q.size();
            result++;
            while (size-- > 0) {
                Node cur = q.poll();
                int y = cur.y;
                int x = cur.x;
                if (y == n -1 && x == m - 1) return result;
                int breakCount = cur.breakCount;
                for (int dir = 0; dir < 4; dir++) {
                    int ny = y + dy[dir];
                    int nx = x + dx[dir];
                    if (ny < 0 || nx < 0 || ny >= n || nx >= m) continue;
                    if (breakCount == 0) {
                        if (!visited[ny][nx][breakCount] && a[ny][nx] == 0) {
                            q.add(new Node(ny, nx, breakCount));
                            visited[ny][nx][breakCount] = true;
                        }
                        if (!visited[ny][nx][breakCount + 1] && a[ny][nx] == 1) {
                            q.add(new Node(ny, nx, breakCount + 1));
                            visited[ny][nx][breakCount + 1] = true;
                        }
                    } else if (breakCount == 1) {
                        if (!visited[ny][nx][breakCount] && a[ny][nx] == 0) {
                            q.add(new Node(ny, nx, breakCount));
                            visited[ny][nx][breakCount] = true;
                        }
                    }
                }
            }
        }

        return -1;
    }

}